Current state Quality Declaration for rcutils (#240)
ros2/rcutils: QUALITY_DECLARATION.md

This document is a declaration of software quality for the `rcutils` package, based on the guidelines in [REP-2004](https://github.com/ros-infrastructure/rep/blob/rep-2004/rep-2004.rst).

# `rcutils` Quality Declaration

The package `rcutils` claims to be in the **Quality Level 4** category.

> Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Requirements for Quality Level 1 in REP-2004 ](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#quality-level-1) of the ROS2 developer guide.

Reviewer: ```suggestion
Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Quality Categories in REP-2004](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#package-quality-categories) of the ROS2 developer guide.
```

## Version Policy [1]

### Version Scheme [1.i]
`rcutils` uses `semver` according to the recommendation for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#versioning)

### Version Stability [1.ii]

> Currently this package version is 0.7.5, soon to be released as 1.0.0

Reviewer 2: How about keeping this declaration completely non-aspirational and removing all the "soon" statements? We know we're planning on getting to it in the coming months, so we can update the entire sentence when we do.

Also, this kind of statement tends to not age well in case we don't manage to get to it. Future readers won't know when "soon" is.

Author: Completed in [f14f23f](https://github.com/ros2/rcutils/pull/240/commits/f14f23f98d2a850a620d40e809f002227627eda1)

### ABI Stability Policy [1.v]

`rcutils` contains C code and therefore must be concerned with ABI stability, and will maintain ABI stability within a ROS distribution.

### ABI and ABI Stability Within a Released ROS Distribution [1.vi]

`rcutils` will not break API nor ABI within a released ROS distribution, i.e. no major releases once the ROS distribution is released.

## Change Control Process [2]

`rcutils` follows the recommended guidelines for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process).

### Change Requests [2.i]
All changes will occur through a pull request, check [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.

### Contributor Origin [2.ii]
All pull requests must have confirmation of contributor origin. This repository uses DCO.

### Peer Review Policy [2.iii]
All pull request will be peer-reviewed, check [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.

## Documentation [3]

### Feature Documentation [3.i]

> `rcutils` does not have a documented feature list. Soon to be added.

Reviewer: I think in the main page of the API live the feature list http://docs.ros2.org/eloquent/api/rcutils/index.html

Author: I think the correct link for the foxy release would be something like "[http://docs.ros2.org/foxy/api/rcutils/index.html](http://docs.ros2.org/foxy/api/rcutils/index.html)". However, that points to an outdated API. A quick check shows it is missing the function `rcutils_strerror`, which appears in the header files but not in that document

### Public API Documentation [3.ii]

> `rcutils` does not cover a public API documentation. Soon to be added.

Reviewer: Public API is available here http://docs.ros2.org/eloquent/api/rcutils/index.html

Author: Same comment as below

### License [3.iii]

The license for `rcutils` is Apache 2.0, and a summary is in each source file, the type is declared in the [`package.xml`](./package.xml) manifest file, and a full copy of the license is in the [`LICENSE`](./LICENSE) file.

There is an automated test which runs a linter that ensures each file has a license statement.

### Copyright Statements [3.iv]

The copyright holders each provide a statement of copyright in each source code file in `rcutils`.

There is an automated test which runs a linter that ensures each file has at least one copyright statement.

## Testing [4]

### Feature Testing [4.i]

Each feature in `rcutils` has corresponding tests which simulate typical usage, and they are located in the [`test`](https://github.com/ros2/rcutils/tree/master/test) directory.
New features are required to have tests before being added.

> 

Reviewer: ```suggestion
Currently nightly results can be seen here:
* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/rcutils/)
* [linux-arm64_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/rcutils/)
* [mac_osx_release](https://ci.ros2.org/view/nightly/job/nightly_osx_release/lastBuild/testReport/rcutils/)
* [windows_release](https://ci.ros2.org/view/nightly/job/nightly_win_rel/lastBuild/testReport/rcutils/)
```

### Public API Testing [4.ii]

Each part of the public API has tests, and new additions or changes to the public API require tests before being added.
The tests aim to cover both typical usage and corner cases, but are quantified by contributing to code coverage.

### Coverage [4.iii]

`rcutils` follows the recommendations for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#code-coverage), and opts to use line coverage instead of branch coverage.

This includes:

- tracking and reporting line coverage statistics
- achieving and maintaining a reasonable branch line coverage (90-100%)
- no lines are manually skipped in coverage calculations

Changes are required to make a best effort to keep or increase coverage before being accepted, but decreases are allowed if properly justified and accepted by reviewers.

> Current coverage statistics can be viewed [here](https://ci.ros2.org/job/ci_linux_coverage/85/cobertura/src_ros2_rcutils_src/)

Reviewer: ```suggestion
Current coverage statistics can be viewed [here](https://ci.ros2.org/job/ci_linux_coverage/lastBuild/cobertura/src_ros2_rcutils_src/)
```

### Performance [4.iv]

`rcutils` follows the recommendations for performance testing of C code in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#performance), and opts to do performance analysis on each release rather than each change.
Soon to be defined if this package requires performance testing and how addresses this topic.

> ### Linters and Static Analysis [4.v]

Reviewer: again links to the night builds?

Author: Added in [efce310](https://github.com/ros2/rcutils/pull/240/commits/efce310a2f85e5b473acb73c29633821fa7d6bd2)

`rcutils` uses and passes all the ROS2 standard linters and static analysis tools for a C package as described in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#linters-and-static-analysis). Passing implies there are no linter/static errors when testing against CI of supported platforms.

## Dependencies [5]

### Direct Runtime ROS Dependencies [5.i]
`rcutils` has no run-time or build-time dependencies that need to be considered for this declaration.

It has several "buildtool" dependencies, which do not affect the resulting quality of the package, because they do not contribute to the public library API.

### Optional Direct Runtime ROS Dependencies [5.ii]
`rcutils` has no run-time or build-time dependencies that need to be considered for this declaration.

### Direct Runtime non-ROS Dependency [5.iii]
`rcutils` has no run-time or build-time dependencies that need to be considered for this declaration.

> ## Platform Support [6]

Reviewer: Links to build status in each platform?

Author: Added in [efce310](https://github.com/ros2/rcutils/pull/240/commits/efce310a2f85e5b473acb73c29633821fa7d6bd2)

Additionally, there are generated header files which are installed and therefore part of the public API.
The source templates for these generated headers are in the `resource` folder.

### API Stability Policy [1.iv]

`rcutils` will not break public API within a released ROS distribution, i.e. no major releases once the ROS distribution is released.

### ABI Stability Policy [1.v]

`rcutils` contains C code and therefore must be concerned with ABI stability, and will maintain ABI stability within a ROS distribution.

### ABI and ABI Stability Within a Released ROS Distribution [1.vi]

`rcutils` will not break API nor ABI within a released ROS distribution, i.e. no major releases once the ROS distribution is released.

## Change Control Process [2]

`rcutils` follows the recommended guidelines for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process).

### Change Requests [2.i]
All changes will occur through a pull request, check [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.

### Contributor Origin [2.ii]

> All pull requests must have confirmation of contributor origin. This repository uses DCO.

Reviewer: ```suggestion
This package uses DCO as its confirmation of contributor origin policy. More information can be found in [CONTRIBUTING](./CONTRIBUTING.md)
```

### Peer Review Policy [2.iii]
All pull request will be peer-reviewed, check [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.

## Documentation [3]

### Feature Documentation [3.i]

`rcutils` does not have a documented feature list.

### Public API Documentation [3.ii]

`rcutils` does not cover a public API documentation.

### License [3.iii]

The license for `rcutils` is Apache 2.0, and a summary is in each source file, the type is declared in the [`package.xml`](./package.xml) manifest file, and a full copy of the license is in the [`LICENSE`](./LICENSE) file.

There is an automated test which runs a linter that ensures each file has a license statement.

### Copyright Statements [3.iv]

The copyright holders each provide a statement of copyright in each source code file in `rcutils`.

> There is an automated test which runs a linter that ensures each file has at least one copyright statement.

Reviewer: link http://build.ros2.org/view/Epr/job/Epr__rcutils__ubuntu_bionic_amd64/lastBuild/testReport/rcutils/copyright/

This document is a declaration of software quality for the `rcutils` package, based on the guidelines in [REP-2004](https://github.com/ros-infrastructure/rep/blob/rep-2004/rep-2004.rst).

# `rcutils` Quality Declaration

The package `rcutils` claims to be in the **Quality Level 4** category.

Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Requirements for Quality Level 1 in REP-2004 ](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#quality-level-1) of the ROS2 developer guide.

> ## Version Policy

Reviewer: In the REP 2004 there is a guideline for Quality Declaration in `markdown` which add the index referencing the REP to the title. Right now looks like:

```
## Version Policy
```
But the guideline define this as
```
## Version Policy [1]
```

I suggest to change all the titles to follow the section number as defined in the REP

> ## Version Policy

Author: I actually added the numbers, but I think I left a repeated section, I will remove it!

### Version Scheme

> (Missing part, bumping to version 1.0.0)

Reviewer: Instead of TODOs or comments in the doc. I suggest to add a column with the QD that we are targeting

Author: I didn't like the idea of the "TODOs" as part of the current status table because it does reflect something aspirational, and not the current state of the quality declaration. But I liked your idea of summarizing it somewhere to avoid browsing the files to look for TODOs. I opened a new section where all those are listed, with the proper index section. Let me know your thoughts!

Reviewer 2: I think for now it's enough that each package has an issue tracking what's missing for level 1. The declaration is meant to declare facts, I think we can leave the aspirational part out of it.

## Change Control Process

`rcutils` follows the recommended guidelines for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process).

This includes:

- all changes occur through a pull request
- all pull request will be peer-reviewed
- all pull request must pass CI on all [tier 1 platforms](https://github.com/ros-infrastructure/rep/blob/master/rep-2000.rst)
- all pull request must resolve related documentation changes before merging

> 

Reviewer: - Must have confirmation of contributor origin. This repository uses DCO.

Reviewer: This will go in the section 2.ii

## Documentation

### Feature Documentation

> TODO fix link (Missing part)

Author: Please let me know your thoughts on the last TO-DO section I added, although is somewhat optional as to the purpose of the declaration of the current state itself.

> TODO fix link (Missing part)

Reviewer 2: I feel that the TODOs are already captured in several places, so we could avoid duplication and remove that section. For example, the TODO for `[3.i]` is already:

* on issue #205
* on the `[3.i]` section itself: i.e. `does not have a documented feature list`
* on the status summary table: i.e. `3.i | Per feature documentation | x`

> TODO fix link (Missing part)

Author: Agreed, removed in [f14f23f](https://github.com/ros2/rcutils/pull/240/commits/f14f23f98d2a850a620d40e809f002227627eda1)

`rcutils` has a [feature list](TODO) and each item in the list links to the corresponding feature documentation.
There is documentation for all of the features, and new features require documentation before being added.

### Public API Documentation

TODO fix link (Missing part)

`rcutils` has embedded API documentation and it is generated using doxygen and is [hosted](TODO) along side the feature documentation.
There is documentation for all of the public API, and new additions to the public API require documentation before being added.

### License

> The license for `rcutils` is Apache 2.0, and a summary is in each source file, the type is declared in the [`package.xml`](https://github.com/ros2/rcutils/blob/master/package.xml) manifest file, and a full copy of the license is in the [`LICENSE`](https://github.com/ros2/rcutils/blob/master/LICENSE) file.

Reviewer: ```suggestion
The license for `rcutils` is Apache 2.0, and a summary is in each source file, the type is declared in the [`package.xml`](./package.xml) manifest file, and a full copy of the license is in the [`LICENSE`](./LICENSE) file.
```

`rcutils` supports all of the tier 1 platforms as described in [REP-2000](https://www.ros.org/reps/rep-2000.html#support-tiers), and tests each change against all of them.

# Current status Summary

The chart below compares the requirements in the REP-2004 with the current state of the rcutils package.
|Number|  Requirement| Current state |
|--|--|--|
|1| **Version policy** |---|
|1.i|Version Policy available | ✓ |
|1.ii|Stable version |☓|
|1.iii|Declared public API|✓|
|1.iv|API stability policy|✓|
|1.v|ABI stability policy|✓|
|1.vi_|API/ABI stable within ros distribution|✓|

> |2| **Change control proces**s |---|

Reviewer: ```suggestion
|2| **Change control process** |---|
```
